node/cmd/spy/spy_test.go

```diff
@@ -0,0 +1,754 @@
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+	publicrpcv1 "github.com/certusone/wormhole/node/pkg/proto/publicrpc/v1"
+	spyv1 "github.com/certusone/wormhole/node/pkg/proto/spy/v1"
+	ipfslog "github.com/ipfs/go-log/v2"
+	"github.com/wormhole-foundation/wormhole/sdk/vaa"
+	"go.uber.org/zap"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+const vaaNonce = uint32(1)
+const vaaSequence = uint64(1)
+const tx = "39c5940431b1507c2a496e945dfb6b6760771fb3c19f2531c5976decc16814ca"
+
+var govEmitter = vaa.Address{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4}
+
+// govAddress is the string representation of the govEmitter Address.
+// Leaving here in case it is helpful for future tests.
+// const govAddress = "0000000000000000000000000000000000000000000000000000000000000004"
+
+func getTxBytes(tx string) []byte {
+	bytes, err := hex.DecodeString(tx)
+	if err != nil {
+		panic("failed decoding Tx string to bytes")
+	}
+	return bytes
+}
+
+// helper method for *vaa.VAA creation
+func getVAA(chainID vaa.ChainID, emitterAddr vaa.Address, nonce uint32) *vaa.VAA {
+	var payload = []byte{97, 97, 97, 97, 97, 97}
+
+	vaa := &vaa.VAA{
+		Version:          vaa.SupportedVAAVersion,
+		GuardianSetIndex: uint32(1),
+		Signatures:       nil,
+		Timestamp:        time.Unix(0, 0),
+		Nonce:            nonce,
+		Sequence:         vaaSequence,
+		ConsistencyLevel: uint8(32),
+		EmitterChain:     chainID,
+		EmitterAddress:   emitterAddr,
+		Payload:          payload,
+	}
+
+	return vaa
+}
+
+// helper method for *vaa.BatchVAA creation
+func getBatchVAA(chainID vaa.ChainID, txID []byte, nonce uint32, emitterAddr vaa.Address) *vaa.BatchVAA {
+	v := getVAA(chainID, emitterAddr, nonce)
+	obs := &vaa.Observation{
+		Index:       uint8(0),
+		Observation: v,
+	}
+	var obsList = []*vaa.Observation{}
+	obsList = append(obsList, obs)
+
+	batchVAA := &vaa.BatchVAA{
+		Version:          vaa.BatchVAAVersion,
+		GuardianSetIndex: uint32(1),
+		Signatures:       nil,
+		Observations:     obsList,
+	}
+
+	return batchVAA
+}
+
+// helper method for *gossipv1.SignedBatchVAAWithQuorum creation
+func getBatchVAAQuorumMessage(chainID vaa.ChainID, txID []byte, nonce uint32, emitterAddr vaa.Address) *gossipv1.SignedBatchVAAWithQuorum {
+	batchVaa := getBatchVAA(chainID, txID, nonce, emitterAddr)
+	vaaBytes, err := batchVaa.Marshal()
+	if err != nil {
+		panic("failed marshaling batchVAA.")
+	}
+
+	msg := &gossipv1.SignedBatchVAAWithQuorum{
+		BatchVaa: vaaBytes,
+		ChainId:  uint32(chainID),
+		TxId:     txID,
+		Nonce:    nonce,
+	}
+	return msg
+}
+
+// wait for the server to establish a client subscription before returning.
+func waitForClientSubscriptionInit(server *spyServer) {
+	server.subsAllVaaMu.Lock()
+	subs := len(server.subsAllVaa)
+	server.subsAllVaaMu.Unlock()
+
+	if subs > 0 {
+		return
+	}
+	waitForClientSubscriptionInit(server)
```

Comment:
Everything I've been able to find suggests that the go compiler doesn't implement recursive tail call optimization so this could overflow the stack.  We should stick to a loop.

Also we should add a `time.Sleep(time.Millisecond)` here to avoid unnecessary contention on the Mutex.

Reply:
Interesting, thanks for bringing that up. I read a little about tail-call optimization, I believe you are correct; loops are the way. I updated this to use a loop with a 10 millisecond sleep.